src/XMakeBuildEngine/Evaluation/Expander.cs

```diff
@@ -1640,71 +1651,134 @@ private static class ItemExpander
-                // expands @(Compile->'%(foo)') to a set of items with the Include value of items in the "Compile" list.
-                if (expressionCapture.Captures != null)
+                IList<Tuple<string, S>> itemsFromCapture;
+                brokeEarlyNonEmpty = ExpandExpressionCapture(expander, expressionCapture, items, elementLocation /* including null items */, options, true, out isTransformExpression, out itemsFromCapture);
+
+                if (brokeEarlyNonEmpty)
                 {
-                    isTransformExpression = true;
+                    return null;
+                }
+
+                result = new List<T>(itemsFromCapture.Count);
+
+                foreach (var itemTuple in itemsFromCapture)
+                {
+                    var itemSpec = itemTuple.Item1;
+                    var originalItem = itemTuple.Item2;
+
+                    if (itemSpec != null && originalItem == null)
+                    {
+                        // We have an itemspec, but no base item
+                        result.Add(itemFactory.CreateItem(itemSpec, elementLocation.File));
+                    }
+                    else if (itemSpec != null && originalItem != null)
+                    {
+                        result.Add(itemSpec.Equals(originalItem.EvaluatedIncludeEscaped)
+                            ? itemFactory.CreateItem(originalItem, elementLocation.File) // itemspec came from direct item reference, no transforms
+                            : itemFactory.CreateItem(itemSpec, originalItem, elementLocation.File)); // itemspec came from a transform and is different from its original item
+                    }
+                    else if (includeNullEntries)
+                    {
+                        // The itemspec is null and the base item doesn't matter
+                        result.Add(null);
+                    }
                 }
 
-                ICollection<S> itemsOfType = items.GetItems(expressionCapture.ItemType);
+                return result;
+            }
+
+            /// <summary>
+            /// Expands an expression capture into a list of items
+            /// If the capture uses a separator, then all the items are concatenated into one string using that separator.
+            /// 
+            /// Returns true if ExpanderOptions.BreakOnNotEmpty was passed, expression was going to be non-empty, and so it broke out early.
+            /// </summary>
+            /// 
+            /// <param name="itemsFromCapture">
+            /// List of items.
+            /// 
+            /// Item1 represents the item string, escaped
+            /// Item2 represents the original item.
+            /// 
+            /// Item1 differs from Item2's string when it is coming from a transform
+            /// 
+            /// </param>
+            internal static bool ExpandExpressionCapture<S>(
```

Comment:
Given an item expression capture (sort of AST for item expressions), it expands it to strings. If it's a transform, then it also provides the original items the new strings came from.

```
<a Include="1;2">
<b include="@(a->'%(Identity)a')">
```
`ExpandExpressionCapture(ItemExpressionCapture("@(a)"))` will return `[("1a", ProjectItem("1")), ("2a", ProjectItem("2"))]`

Alternative names:
`ExpandItemExpressionCaptureIntoStringsWithOriginatingItems`
`ExpandItemExpressionCaptureIntoTupleList`
`ExpandItemExpressionCaptureIntoStringAndOriginalItemList`